From c0ca3d70e8d3cf81e2255a217f7ca402f5ed0862 Mon Sep 17 00:00:00 2001
From: Vivek Goyal <vgoyal@redhat.com>
Date: Fri Jul  1 16:34:27 2016 -0400
Subject: [PATCH] ovl: modify ovl_permission() to do checks on two inodes
Git-commit: c0ca3d70e8d3cf81e2255a217f7ca402f5ed0862
References: CVE-2018-16597, bsc#1106512, bsc#1107299
Patch-mainline: v4.8-rc1

Right now ovl_permission() calls __inode_permission(realinode), to do
permission checks on real inode and no checks are done on overlay inode.

Modify it to do checks both on overlay inode as well as underlying inode.
Checks on overlay inode will be done with the creds of calling task while
checks on underlying inode will be done with the creds of mounter.

Signed-off-by: Vivek Goyal <vgoyal@redhat.com>
Signed-off-by: Miklos Szeredi <mszeredi@redhat.com>
Acked-by: Goldwyn Rodrigues <rgoldwyn@suse.com>

---
 fs/overlayfs/inode.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/fs/overlayfs/inode.c
+++ b/fs/overlayfs/inode.c
@@ -10,6 +10,7 @@
 #include <linux/fs.h>
 #include <linux/slab.h>
 #include <linux/xattr.h>
+#include <linux/cred.h>
 #include "overlayfs.h"
 
 static int ovl_copy_up_truncate(struct dentry *dentry)
@@ -93,6 +94,7 @@ int ovl_permission(struct inode *inode,
 	struct inode *realinode;
 	struct dentry *realdentry;
 	bool is_upper;
+	const struct cred *old_cred;
 	int err;
 
 	if (S_ISDIR(inode->i_mode)) {
@@ -168,7 +170,17 @@ int ovl_permission(struct inode *inode,
 			goto out_dput;
 	}
 
+	/*
+	 * Check overlay inode with the creds of task and underlying inode
+	 * with creds of mounter
+	 */
+	err = generic_permission(inode, mask);
+	if (err)
+		goto out_dput;
+
+	old_cred = ovl_override_creds(inode->i_sb);
 	err = __inode_permission(realinode, mask);
+	revert_creds(old_cred);
 out_dput:
 	dput(alias);
 	return err;
